pkg/maglev/maglev.go

```diff
@@ -98,10 +95,18 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
 					permutation[i*int(m)+int(j)] = (permutation[i*int(m)+int(j-1)] + skip) % m
 				}
 			}
-			wg.Done()
-		}(g)
+			return nil
+		})
+	}
+	_, err := wp.Drain()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error waiting for permutation to complete: %v\n", err)
```

Comment:
if these fail we'll return bogus `permutation` and somewhat silently break maglev. It would be better to instead refactor these methods to pass the error upwards and handle the error properly. Errors should be logged using the logger and not with Fprintf.

Pinging @brb as this change seems to be a bit more involved.

Reply:
@joamaki Note that `Drain()` actually [returns results for each task](https://pkg.go.dev/github.com/cilium/workerpool#WorkerPool.Drain) so if it's important to report each task error, then the code should iterate over [task results](https://pkg.go.dev/github.com/cilium/workerpool#Task) to collect individual errors. It would then also be advisable to actually use a task ID in `Submit()` rather than the empty string to that one can associate a task error with a specific task.

Reply:
Looking at the implementation looks like these aren't really expected to fail unexpectedly. It would probably make sense to `Fatalf` on errors from workerpool and not bubble them up.

Reply:
thanks very much，i will modify code after the holiday, about one week:) sry!

Reply:
Is it correct to assume that if the callback fn doesn't return any error (except for nil), then `Drain()` is guaranteed to return nil as err? Asking, as it's quite critical for us not to error here, as the service manager is not able to reinsert if a service if any of its provisioning steps has failed.

Reply:
@brb `Drain` returns an error under if one of the following condition is true:

- The workerpool is already closed when `Drain` is called.
- `Drain` is called concurrently.

Errors from tasks don't affect the error returned by `Drain`, callers need iterate over the task results and check for errors if one can be expected. Therefore, I think it's safe to ignore the error here (or just print something in case the code path is modified in a way that one of the above could occur).